### DISCUSSION TITLE
Client-side routing for static site

### DISCUSSION THREAD

**Discussion by pcardoso** [piku/piku#366] [Q&A]
I'm trying to use Piku for a project written in React that uses client-side routing. I'm deploying it as a static site.

Things work, the URL is updated as I navigate but when I hit reload or directly enter a non-root path I get a 404.

Typically for these kind of projects we need to serve the root index.html page for all path requests, as in 

`try_files $uri /index.html;` [1] 

Looking at the app.conf nginx configuration file in `/home/piku/.piku/nginx` for the app I see the try_files directive as 

`try_files $uri $uri.html $uri/ =404;`

So it won't serve index.html when the path is not found.

Using a custom nginx.conf file didn't work for me, I tried writing a new `location /` directive but nginx complains about duplicate sections.

Using a small server-side shim to serve the files or using hash-routing would work but both are quite inelegant.

Is there a way to make this work? 

Thanks

[1] https://stackoverflow.com/questions/43951720/react-router-and-nginx

**Comment by chr15m** [comment]
Sounds like a useful feature and quite simple to implement. What if there was an ENV variable like `NGINX_CATCHALL`? I guess this could be a separately applied block so you could set it to an arbitrary route and it would catch everything below that.

**Reply by pcardoso** [reply]
Taking your suggestion, I did the following modifications:

my ENV file: `NGINX_CATCH_ALL=index.html`

in piku.py, modified PIKU_INTERNAL_NGINX_STATIC_MAPPING line to 

`try_files $uri $uri.html $uri/ $catch_all =404;`

and added this to line 938, where the static worker is handled.

`catch_all = env.get('NGINX_CATCH_ALL', '')`

This makes the catch_all available in locals(). 

This works for me but may not be the right way to do this. 😅

**Reply by chr15m** [reply]
That's great and a nice minimal implementation. Would you be up for raising a pull request?

**Reply by pcardoso** [reply]
Sure! Glad I can help!

**Reply by pcardoso** [reply]
PR here, should I also add an example?

https://github.com/piku/piku/pull/367

**Reply by chr15m** [reply]
I think the updated docs are sufficient thanks!
